src/libexpr/print.cc

```diff
@@ -90,5 +119,286 @@ printAttributeName(std::ostream & str, std::string_view name) {
+        return std::forward_as_tuple(!lhsIsImportant, lhs)
+            < std::forward_as_tuple(!rhsIsImportant, rhs);
+    }
+};
+
+typedef std::set<Value *> ValuesSeen;
+
+class Printer
+{
+    private:
+        EvalState & state;
+        PrintOptions options;
+
+        void print(std::ostream & str, Value & v, size_t depth, ValuesSeen * seen);
+
+    public:
+        Printer(EvalState & s, PrintOptions o): state(s), options(o) { }
+        void print(std::ostream & str, Value & v)
+        {
+            ValuesSeen seen;
+            print(str, v, 0, options.trackRepeated ? &seen : nullptr);
+        }
+};
+
+void Printer::print(std::ostream & str, Value & v, size_t depth, ValuesSeen * seen)
+{
+    str.flush();
+    checkInterrupt();
+
+    if (options.force) {
+        try {
+            state.forceValue(v, [&]() { return v.determinePos(noPos); });
+        } catch (AssertionError & e) {
+            if (options.ansiColors)
+                str << ANSI_RED;
+            str << "«" << e.msg() << "»";
+            if (options.ansiColors)
+                str << ANSI_NORMAL;
+            return;
+        }
+    }
+
+    switch (v.type()) {
+
+    case nInt:
+        if (options.ansiColors)
+            str << ANSI_CYAN;
+        str << v.integer;
+        if (options.ansiColors)
+            str << ANSI_NORMAL;
+        break;
+
+    case nFloat:
+        if (options.ansiColors)
+            str << ANSI_CYAN;
+        str << v.fpoint;
+        if (options.ansiColors)
+            str << ANSI_NORMAL;
+        break;
+
+    case nBool:
+        if (options.ansiColors)
+            str << ANSI_CYAN;
+        printLiteralBool(str, v.boolean);
+        if (options.ansiColors)
+            str << ANSI_NORMAL;
+        break;
+
+    case nString:
+        printLiteralString(str, v.string_view(), options.maxStringLength, options.ansiColors);
+        break;
+
+    case nPath:
+        if (options.ansiColors)
+            str << ANSI_GREEN;
+        str << v.path().to_string(); // !!! escaping?
+        if (options.ansiColors)
+            str << ANSI_NORMAL;
+        break;
+
+    case nNull:
+        if (options.ansiColors)
+            str << ANSI_CYAN;
+        str << "null";
+        if (options.ansiColors)
+            str << ANSI_NORMAL;
+        break;
+
+    case nAttrs: {
+        if (seen)
+            seen->insert(&v);
+
+        if (options.force && state.isDerivation(v)) {
+            str << "«derivation";
+            Bindings::iterator i = v.attrs->find(state.sDrvPath);
+            NixStringContext context;
+            if (i != v.attrs->end())
+                str << " " << state.store->printStorePath(state.coerceToStorePath(i->pos, *i->value, context, "while evaluating the drvPath of a derivation"));
+            str << "»";
+        }
+
+        else if (depth < options.maxDepth) {
+            str << "{ ";
+
+            typedef std::map<std::string, Value *, ImportantFirstAttrNameCmp> Sorted;
```

Comment:
Could you make the importance sorting conditional on whether `maxAttrs` has been restricted? Perhaps by making `sorted` a `vector`, and then choosing between different sort calls.

Reply:
If we keep the old printing function for `nix-instantiate`, it may be ok to always prioritize these attribute names. (Although this was debated in https://github.com/NixOS/nix/pull/9606#discussion_r1432043993, because of `attrNames`, I'd be ok with pulling important info to the front unconditionally.)